### PR TITLE
ARM ASM: ARMv7a with NEON instructions

### PR DESCRIPTION
**Pull request by SparkiDev**
# Description

Change to build assembly code for ARMv7a with NEON instruction set.
./configure -host=armv7a --enable-armasm
Added ARM32 SHA-256 NEON only implementation.

# Testing

Tested SHA-256 NEON implementation using QEMU and Raspberry Pi 3 (ARMv7 Processor rev 4 (v7l))

```
./configure -host=armv7a --enable-armasm --enable-debug --disable-shared CFLAGS="-fomit-frame-pointer -DWOLFSSL_ARM_ARCH_NEON_64BIT" && make check
./configure -host=armv7a --enable-armasm CFLAGS="-DWOLFSSL_ARM_ARCH_NEON_64BIT" && make check
```

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Comment by SparkiDev** [comment]
retest this please

**Comment by douzzer** [comment]
@SparkiDev I've just done some testing on this.

First thing I had to change the `armv7a` to `armv7a*` in configure.ac L2050, to accommodate the standard target tuple `armv7a-unknown-linux-gnueabihf`.

That done, to get `--enable-armasm` to build, I had to include `--disable-chacha --disable-xchacha --disable-poly1305`, because of errors like this:
```
/tmp/tmp.4346_8721/ccFSpc8A.s: Assembler messages:
/tmp/tmp.4346_8721/ccFSpc8A.s:270: Error: first transfer register must be even -- `ldrd r11,r10,[r14,#4*14]'
make[2]: *** [Makefile:5938: wolfcrypt/src/port/arm/src_libwolfssl_la-armv8-chacha.lo] Error 1
```

Once I did that, `testsuite.test` nominally succeeded, but crashed at exit:
```
[...]
mutex    test passed!
memcb    test passed!
Test complete
qemu: uncaught target signal 7 (Bus error) - core dumped
```
(This was repeatable.)


Other bits:

With `-pedantic`, this happens:
```
wolfcrypt/src/sha256.c:2002: error: ISO C forbids an empty translation unit [-Werror=pedantic]
```
Looks like an include.am oversight.

And finally, `--enable-all --enable-armasm` on armv7a breaks as follows:
```
wolfcrypt/src/port/arm/armv8-sha256.c: In function ‘wc_Sha256Transform’:
089ebf277f (<sean@wolfssl.com> 2021-03-23 12:53:06 +1000 1540)     Sha256Transform(sha256, data, 1);
wolfcrypt/src/port/arm/armv8-sha256.c:1540:5: error: implicit declaration of function ‘Sha256Transform’; did you mean ‘wc_Sha256Transform’? [-Werror=implicit-function-declaration]
 1540 |     Sha256Transform(sha256, data, 1);
      |     ^~~~~~~~~~~~~~~
      |     wc_Sha256Transform
089ebf277f (<sean@wolfssl.com> 2021-03-23 12:53:06 +1000 1540)     Sha256Transform(sha256, data, 1);
wolfcrypt/src/port/arm/armv8-sha256.c:1540:5: error: nested extern declaration of ‘Sha256Transform’ [-Werror=nested-externs]
```
This just _has_ to be related to the wholesale replacement of sha256.c that happens for armasm.

The toolchain versions for the above:
```
[I-O] [  ] cross-armv7a-unknown-linux-gnueabihf/binutils-2.38-r2:2.38
[I-O] [  ] cross-armv7a-unknown-linux-gnueabihf/gcc-12.1.1_p20220625:12
[I-O] [  ] cross-armv7a-unknown-linux-gnueabihf/linux-headers-5.19:0
[I-O] [  ] cross-armv7a-unknown-linux-gnueabihf/glibc-2.35-r8:2.2
[IP-] [  ] app-emulation/qemu-7.0.0-r3:0
```

**Comment by haydenroche5** [comment]
Had to disable the same things as @douzzer. But, I was able to get some benchmarking results on my Raspberry Pi:

```
pi@raspberrypi:~/wolfssl $ lscpu; cat before_sha256.txt; cat after_sha256.txt; cat before_sha512.txt; cat after_sha512.txt 
Architecture:        armv7l
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  1
Core(s) per socket:  4
Socket(s):           1
Vendor ID:           ARM
Model:               3
Model name:          Cortex-A72
Stepping:            r0p3
CPU max MHz:         1500.0000
CPU min MHz:         600.0000
BogoMIPS:            108.00
Flags:               half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
------------------------------------------------------------------------------
 wolfSSL version 5.4.0
------------------------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
SHA-256             80 MB took 1.008 seconds,   79.339 MB/s
Benchmark complete
------------------------------------------------------------------------------
 wolfSSL version 5.4.0
------------------------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
SHA-256            120 MB took 1.020 seconds,  117.677 MB/s
Benchmark complete
------------------------------------------------------------------------------
 wolfSSL version 5.4.0
------------------------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
SHA-512             40 MB took 1.019 seconds,   39.273 MB/s
Benchmark complete
------------------------------------------------------------------------------
 wolfSSL version 5.4.0
------------------------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
SHA-512             75 MB took 1.064 seconds,   70.472 MB/s
Benchmark complete
```

Looking quite a bit faster!

**Comment by douzzer** [comment]
just tried this rebased on latest master, with the configure.ac tweak to recognize `armv7a-unknown-linux-gnueabihf`:
```[cross-armv7a-all-armasm] [1 of 1] [08b4fd6f2e-dirty]
    autogen.sh 08b4fd6f2e-dirty...   real 0m14.917s  user 0m13.195s  sys 0m0.859s
    configure...   real 0m8.970s  user 0m4.785s  sys 0m5.044s
    build...   real 0m30.472s  user 2m5.040s  sys 0m6.234s
    testsuite.test...   real 0m28.347s  user 0m28.304s  sys 0m0.014s
testsuite.test for scenario cross-armv7a-all-armasm exited with status 135.
================================================================================
TLSv1.3 KDF test passed!
X963-KDF    test passed!
GMAC     test passed!
ARC4     test passed!
DES      test passed!
DES3     test passed!
AES      test passed!
AES192   test passed!
AES256   test passed!
AESOFB   test passed!
AES-GCM  test passed!
AES-CCM  test passed!
AES Key Wrap test passed!
AES-SIV  test passed!
CAMELLIA test passed!
RSA NOPAD test passed!
RSA      test passed!
DH       test passed!
DSA      test passed!
qemu: uncaught target signal 7 (Bus error) - core dumped
[... retries ...]
    cross-armv7a-all-armasm fail_check
    failed config: '--enable-all' 'CPPFLAGS=-DNO_WOLFSSL_CIPHER_SUITE_TEST -DWOLFSSL_OLD_PRIME_CHECK -pedantic' '--enable-asn=template' '--enable-armasm' '--disable-chacha' '--disable-poly1305' '--disable-xchacha' '--host=armv7a-unknown-linux-gnueabihf' 'FILECMD=/bin/false' 'MANIFEST_TOOL=/bin/false'
```

**Comment by SparkiDev** [comment]
@douzzer

Can you please try without SRP (--disable-srp).
I'm not seeing any issue with QEMU on Linux.
And also try to enable ChaCha, Poly1305 and XChaCha.

Thanks,
Sean

**Comment by paulwratt** [comment]
shouldn't `-mfpu=neon-vfpv3` just be plain `-mfpu=neon` or `-mfpu=neon-vfp`, `neon` being a generic Arm processor, `neon-vfp` being certain types of Arm processor, and `neon-vfpv3` being specific Arm processor (which exludes v1, v2 & v4)?

**Comment by SparkiDev** [comment]
Hi @paulwratt,

I have found it difficult in the past to get the right mfpu setting that will work!
Using 'neon' works for me with QEMU so I've changed configure.ac.
Note that the appropriate mfpu setting should be decided by each customer based on their hardware. 

Thanks,
Sean

**Comment by haydenroche5** [comment]
Tested with `./configure --host=armv7a --enable-armasm` on Raspberry Pi. No build errors, now, but testwolfcrypt is failing:

```
pi@raspberrypi:~/wolfssl $ ./wolfcrypt/test/testwolfcrypt 
------------------------------------------------------------------------------
 wolfSSL version 5.5.0
------------------------------------------------------------------------------
error    test passed!
MEMORY   test passed!
base64   test passed!
asn      test passed!
RANDOM   test passed!
MD5      test passed!
SHA      test passed!
SHA-256  test passed!
SHA-384  test passed!
SHA-512  test passed!
Hash     test passed!
HMAC-MD5 test passed!
HMAC-SHA test passed!
HMAC-SHA256 test passed!
HMAC-SHA384 test passed!
HMAC-SHA512 test passed!
HMAC-KDF    test passed!
TLSv1.3 KDF test passed!
GMAC     test passed!
Chacha   test failed!
 error = -4726
Exiting main with return code: -1
```

**Comment by SparkiDev** [comment]
@haydenroche5 
The error number is not unique.
I've updated the test and made minor changes to the ChaCha asm.
When you have time, run it again so I can determine which test case is actually failing.

Thanks,
Sean

**Comment by SparkiDev** [comment]
By the way, which model of the Pi do you have?
May help to know which CPU you are using.

**Comment by SparkiDev** [comment]
@douzzer 
Do you know which CPU is being emulated?

**Comment by haydenroche5** [comment]
> When you have time, run it again so I can determine which test case is actually failing.

Everything passed with the latest changes. :)

> By the way, which model of the Pi do you have?

```
pi@raspberrypi:~ $ lscpu
Architecture:        armv7l
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  1
Core(s) per socket:  4
Socket(s):           1
Vendor ID:           ARM
Model:               3
Model name:          Cortex-A72
Stepping:            r0p3
CPU max MHz:         1500.0000
CPU min MHz:         600.0000
BogoMIPS:            108.00
Flags:               half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
```

**Comment by SparkiDev** [comment]
Bizarre! I changed the code to use simpler instructions and that was the difference!
Which version of the Raspberry Pi is it? v4 from the specs.

**Comment by paulwratt** [comment]
> Which version of the Raspberry Pi is it? v4 from the specs

yeah `vfpv4` signifies RPi4

> Using 'neon' works for me with QEMU so I've changed configure.ac

Thanks, that should get non-RPi armv7 to compile with neon instructions as well.

FYI older RPi (that latest 2B+ & 3) only have `vfp` and `vfpv2` (thats the model 2B & older 2B+). All RPi have `vfp`. Non-RPi will only have `vfp` if its a Broadcom VC based Arm CPU. `neon` works for any Arm processor with a _compliant_ GPU

File: src/include.am [hunk 1]
```diff
@@ -203,6 +203,11 @@ endif
 
 if BUILD_ARMASM
 src_libwolfssl_la_SOURCES += wolfcrypt/src/port/arm/armv8-sha256.c
+if BUILD_ARMASM_INLINE
```
Comment (paulwratt) [inline src/include.am:206]: this builds the C inline, else builds Asm

Is this correct?

File: src/include.am [hunk 2]
```diff
@@ -300,10 +305,15 @@ endif
 endif !BUILD_FIPS_CURRENT
 
 if !BUILD_FIPS_CURRENT
+src_libwolfssl_la_SOURCES += wolfcrypt/src/sha256.c
 if BUILD_ARMASM
 src_libwolfssl_la_SOURCES += wolfcrypt/src/port/arm/armv8-sha256.c
+if BUILD_ARMASM_INLINE
```
Comment (paulwratt) [inline src/include.am:311]: same again, and if its not defined it will also build build the Asm

Is this correct?

**Comment by paulwratt** [comment]
I did not use the "Begin Review", but wanted to see what sort of "simplified assember instruction" you were talking about, but came accros these:
`if BUILD_ARMASM_INLINE`
which seemed to be flipped (unless I dont correctly evaluate Make use of that statement)

**Comment by haydenroche5** [comment]
> > Which version of the Raspberry Pi is it? v4 from the specs
> 
> yeah `vfpv4` signifies RPi4

Indeed. RPi 4 Model B.

File: src/include.am (hunk 1 above)
Comment (SparkiDev) [reply src/include.am:206]: Yes. I generate the assembly code as inline C and assembly code.
The files provide the same functions with the same instructions.

File: src/include.am (hunk 2 above)
Comment (SparkiDev) [reply src/include.am:311]: Yes, as above.

**Comment by SparkiDev** [comment]
BUILD_ARMASM_INLINE means the assembly code is inlined in the C code file.
There is no difference between the two files except that one can be compiled with a C compiler and the other needs go through the assembler.

WOLFSSL_ARMASM_NO_CRYPTO means the CPU doesn't support cryptographic instructions.
Therefore assembly implementations not using cryptographic instructions have been introduced.
Given the obvious strangeness of the define name I've changed it to WOLFSSL_ARMASM_NO_HW_CRYPTO.

**Comment by SparkiDev** [comment]
@douzzer
Could you try again?
I've changed the -mfpu option and it might work now.

Thanks,
Sean

**Review by dgarske — CHANGES_REQUESTED** [review]
On commit: 5dc1732036ddcee4f9ad353f735a3dd2defc5908

```
cat /proc/cpuinfo
processor	: 0
model name	: ARMv7 Processor rev 4 (v7l)
BogoMIPS	: 83.20
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xd03
CPU revision	: 4
...
Hardware	: BCM2835
Revision	: a22082
Serial		: 000000000920b30e
```

Getting "bus errors" on my Raspberry Pi 3.

```
./configure -host=armv7a --enable-armasm
...
configure: 32bit ARMv7-a found, setting mfpu to neon
...
./testsuite/testsuite.test
------------------------------------------------------------------------------
 wolfSSL version 5.5.0
------------------------------------------------------------------------------
error    test passed!
MEMORY   test passed!
base64   test passed!
asn      test passed!
RANDOM   test passed!
MD5      test passed!
SHA      test passed!
SHA-256  test passed!
SHA-384  test passed!
SHA-512  test passed!
Hash     test passed!
HMAC-MD5 test passed!
HMAC-SHA test passed!
HMAC-SHA256 test passed!
HMAC-SHA384 test passed!
HMAC-SHA512 test passed!
HMAC-KDF    test passed!
TLSv1.3 KDF test passed!
GMAC     test passed!
Chacha   test passed!
POLY1305 test passed!
ChaCha20-Poly1305 AEAD test passed!
AES      test passed!
AES192   test passed!
AES256   test passed!
AES-GCM  test passed!
RSA      test passed!
DH       test passed!
PWDBASED test passed!
ECC      test passed!
logging  test passed!
time test passed!
mutex    test passed!
memcb    test passed!
Test complete
Bus error
```

```
./configure -host=armv7a --enable-armasm --enable-debug --disable-shared CFLAGS="-fomit-frame-pointer" && make
...
configure: 32bit ARMv7-a found, setting mfpu to neon
...
gdb ./testsuite/testsuite.test
...
Shrinking input buffer
growing output buffer

Program received signal SIGBUS, Bus error.
[Switching to Thread 0x76dbf450 (LWP 2459)]
Transform_Sha256_Len () at wolfcrypt/src/port/arm/armv8-32-sha256-asm.S:1562
1562		vrev32.8	q0, q0
(gdb)
(gdb) bt
#0  Transform_Sha256_Len () at wolfcrypt/src/port/arm/armv8-32-sha256-asm.S:1562
#1  0x00000000 in ?? ()
```

**Comment by SparkiDev** [comment]
Not all ARMv7a CPUs support all NEON instructions.
Define WOLFSSL_ARM_ARCH_NO_VREV when vrev not available.
I removed usage of VTRN in a previous commit.

**Review by dgarske — CHANGES_REQUESTED** [review]
Now it is unhappy with `vshl` instruction.

```
./configure -host=armv7a --enable-armasm --enable-debug --disable-shared CFLAGS="-fomit-frame-pointer -DWOLFSSL_ARM_ARCH_NO_VREV" && make
...
./testsuite/testsuite.test
...
Shrinking input buffer
growing output buffer

Program received signal SIGBUS, Bus error.
[Switching to Thread 0x76dbf450 (LWP 7210)]
Transform_Sha256_Len () at wolfcrypt/src/port/arm/armv8-32-sha256-asm.S:1568
1568		vshl.i16	q4, q0, #8
(gdb) bt
#0  Transform_Sha256_Len () at wolfcrypt/src/port/arm/armv8-32-sha256-asm.S:1568
#1  0x00000000 in ?? ()
(gdb)
```

Also please document the build option. Suggest something like this patch:

```
diff --git a/configure.ac b/configure.ac
index a8cc754f2..969ba46f1 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2081,6 +2081,8 @@ then
             AC_MSG_NOTICE([64bit ARMv8 found, setting mcpu to generic+crypto])
             ;;
         armv7a)
+            # Note: Not all ARMv7a Neon CPU's have the "vrev32.8" instruction.
+            #   If seeing a sigbus error on "vrev" use then build with "WOLFSSL_ARM_ARCH_NO_VREV"
             AM_CPPFLAGS="$AM_CPPFLAGS -march=armv7-a -mfpu=neon -DWOLFSSL_ARMASM_NO_HW_CRYPTO -DWOLFSSL_ARM_ARCH=7"
             # Include options.h
             AM_CCASFLAGS="$AM_CCASFLAGS -DEXTERNAL_OPTS_OPENVPN"
```

**Comment by paulwratt** [comment]
@SparkiDev thanks for clarification .. you are almost there by the looks of it

BTW `vfp4` works for the CM4 too (yes?)

**Comment by SparkiDev** [comment]
Found a reference saying that Cortex-A9 and older CPUs only support 64-bit registers and not 128.
Use WOLFSSL_ARM_ARCH_NEON_64BIT to indicate this.
(Don't use VREV define anymore.)

**Comment by dgarske** [comment]
@SparkiDev pulled latest PR 5152 I am still seeing the same issues:

```
./configure -host=armv7a --enable-armasm --enable-debug --disable-shared CFLAGS="-fomit-frame-pointer -DWOLFSSL_ARM_ARCH_NEON_64BIT" && make
...
gdb ./tests/unit.test
...
Program received signal SIGBUS, Bus error.
[Switching to Thread 0x76dbf440 (LWP 6759)]
Transform_Sha256_Len () at wolfcrypt/src/port/arm/armv8-32-sha256-asm.S:1563
1563		vrev32.8	q0, q0
```

```
./configure -host=armv7a --enable-armasm --enable-debug --disable-shared CFLAGS="-fomit-frame-pointer -DWOLFSSL_ARM_ARCH_NEON_64BIT -DWOLFSSL_ARM_ARCH_NO_VREV" && make
...
gdb ./tests/unit.test
...
Program received signal SIGBUS, Bus error.
[Switching to Thread 0x76dbf440 (LWP 12093)]
Transform_Sha256_Len () at wolfcrypt/src/port/arm/armv8-32-sha256-asm.S:1568
1568		vshl.i16	q4, q0, #8
```

**Comment by SparkiDev** [comment]
The define probably isn't reaching the assembly code!

**Comment by dgarske** [comment]
> The define probably isn't reaching the assembly code!

Good point. Trying again with ASFLAGS

**Comment by dgarske** [comment]
Still not working for me. Any suggestions?

```
./configure -host=armv7a --enable-armasm --enable-debug --disable-shared AM_CFLAGS="-fomit-frame-pointer -DWOLFSSL_ARM_ARCH_NEON_64BIT" AM_CCASFLAGS="-fomit-frame-pointer -DWOLFSSL_ARM_ARCH_NEON_64BIT" && make
...
wolfcrypt/src/port/arm/armv8-chacha.c: In function 'wc_Chacha_encrypt_256':
wolfcrypt/src/port/arm/armv8-chacha.c:1383:1: error: fp cannot be used in asm here
 }
 ^
  CC       wolfcrypt/src/src_libwolfssl_la-chacha20_poly1305.lo
```

```
./configure -host=armv7a --enable-armasm AM_CFLAGS="-DWOLFSSL_ARM_ARCH_NEON_64BIT" AM_CCASFLAGS="-DWOLFSSL_ARM_ARCH_NEON_64BIT" && make
 $ ./tests/unit.test
starting unit tests...
 Begin API Tests
   wolfSSL_Init(): passed
   test_wolfSSL_ERR_strings: passed
   wolfSSL_CTX_use_certificate_buffer(): passed
In verification callback, error = 0, unknown error number
	Peer certs: 1
	Subject's domain name at 0 is www.wolfssl.com
In verification callback, error = -188, ASN no signer error to confirm failure
	Peer certs: 1
	Subject's domain name at 0 is www.wolfssl.com
	Allowing failed certificate check, testing only (shouldn't do this in production)
   test_CertRsaPss: passed
Bus error
```

```
git log
commit 2c4c7ba6dad7b286fb14e9cf37c6bfec02f0d890
Author: Sean Parkinson <sean@wolfssl.com>
Date:   Mon Sep 12 10:00:18 2022 +1000

    ARM v7a ASM: 128-bit registers not supported

    Cortex-A5 - Cortex-A9 only support 64-bit wide NEON.
    Remove use of WOLFSSL_ARM_ARCH_NO_VREV.
    Use WOLFSSL_ARM_ARCH_NEON_64BIT to indicate to use 64-bit NEON registers
    and not 128-bit NEON registers.
```

**Comment by SparkiDev** [comment]
Try without debug for now.

**Comment by dgarske** [comment]
> Try without debug for now.

I did. It was in the log. Still his fault

**Comment by dgarske** [comment]
@SparkiDev I pulled latest. Still see the same bus error without debug. With debug same "fp cannot be used" error.

Tested using:
`./configure -host=armv7a --enable-armasm AM_CFLAGS="-DWOLFSSL_ARM_ARCH_NEON_64BIT" AM_CCASFLAGS="-DWOLFSSL_ARM_ARCH_NEON_64BIT" && make`. 
AND
`./configure -host=armv7a --enable-armasm --enable-debug --disable-shared AM_CFLAGS="-fomit-frame-pointer -DWOLFSSL_ARM_ARCH_NEON_64BIT" AM_CCASFLAGS="-fomit-frame-pointer -DWOLFSSL_ARM_ARCH_NEON_64BIT" && make`

**Comment by SparkiDev** [comment]
@dgarske 
Can you tell be which file the compiler is complaining about with the use of fp?
I've eliminated the error with the my QEMU build when DEBUG is defined and not NDEBUG in armv8-chacha.c.

**Comment by dgarske** [comment]
> @dgarske
> Can you tell be which file the compiler is complaining about with the use of fp?
> I've eliminated the error with the my QEMU build when DEBUG is defined and not NDEBUG in armv8-chacha.c.

It was `armv8-chacha.c:1383`. See my comments above.

**Comment by dgarske** [comment]
@SparkiDev 

Still getting a bus error on vrev32.8. The wolfCrypt tests all pass , just the unit test fails.

```
./configure '-host=armv7a' '--enable-armasm' '--enable-debug' '--disable-shared' 'CFLAGS=-fomit-frame-pointer -DWOLFSSL_ARM_ARCH_NEON_64BIT' && make

gdb ./tests/unit.test
Supported Versions extension to write

Program received signal SIGBUS, Bus error.
[Switching to Thread 0x76dbf440 (LWP 5500)]
Transform_Sha256_Len () at wolfcrypt/src/port/arm/armv8-32-sha256-asm.S:1568
1568		vrev32.8	d0, d0
(gdb) bt
#0  Transform_Sha256_Len () at wolfcrypt/src/port/arm/armv8-32-sha256-asm.S:1568
#1  0x00000000 in ?? ()
```

Tests on the Raspberry Pi 3:

```
cat /proc/cpuinfo
processor	: 0
model name	: ARMv7 Processor rev 4 (v7l)
BogoMIPS	: 83.20
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xd03
CPU revision	: 4
...
Hardware	: BCM2835
Revision	: a22082
Serial		: 000000000920b30e
```

**Review by dgarske — APPROVED** [review]
The alignment fixes for SHA-256 and SHA-512 resolve the issues! Wonderful work on this.

**Pull request merged**